### PR TITLE
fix: always use remote token list for default petnames

### PR DESCRIPTION
**Pull request by matthewwalsh0**
## **Description**

Petnames (and by extension transaction simulations) rely on the token list to display default token names and symbols.

With the `Autodetect tokens` preference disabled, static data for ~600 tokens is available on mainnet and no token data is available on other chains. When enabled, data for ~4000 tokens is available on mainnet.

In order to provide the most default token names possible, this PR forces the use of the remote token list for petnames, and therefore enables token list polling if any of the following preferences are enabled:

- `Autodetect tokens` (Token Detection)
- `Allow nicknames` (Petnames)
- `Estimate balance changes` (Transaction Simulations)

To ensure the necessary privacy control, the `Allow nicknames` toggle has also been added to the onboarding settings.

At the technical level, this includes:

- Removing all `TokenListController` logic from the `PreferencesController` to keep it decoupled from individual domain.
- Adding an equivalent listener in the `MetamaskController`.
- Adding an optional `forceRemote` argument to the `getTokenList` selector.
- Using the above argument in the `useDisplayName` hook.
- Updating impacted unit and E2E tests.

[Open in GitHub Codespaces](https://codespaces.new/MetaMask/metamask-extension/pull/23919?quickstart=1)

## **Related issues**

Fixes: [#2339](https://github.com/MetaMask/MetaMask-planning/issues/2339)

### Remote Token List 

1. Disable `Autodetect tokens`.
2. Disable `Estimate balance changes`.
3. Enable `Allow nicknames`.
4. Add Optimism network.
5. Switch to Optimism network.
6. Go to test dApp.
7. Use send form to create transaction targeting Tether USD (`0x94b008aa00579c1307b0ef2c499ad98a8ce58e58`).
10. Verify the default token name is displayed in the petnames component.
11. Repeat with permutations of the three preferences and network.

### Onboarding Setting

1. Install fresh extension.
2. Go to onboarding settings.
3. Disable `Allow nicknames`.
4. Verify nicknames are disabled in experimental settings.

### **Before**

Screenshot 2024-04-09 at 22 15 53

### **After**

Screenshot 2024-04-09 at 22 12 41

Screenshot 2024-04-10 at 20 50 49

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

File: app/scripts/metamask-controller.js [hunk 1]
```diff
@@ -2515,25 +2528,11 @@ export default class MetamaskController extends EventEmitter {
...
-      // TODO: Remove once the preferences controller has been replaced with the core monorepo implementation
-      this.controllerMessenger.publish(
-        'PreferencesController:stateChange',
-        state,
-        [],
-      );
-    });
+    this.preferencesController.store.subscribe(
+      previousValueComparator((prevState, currState) => {
+        this.#onPreferencesControllerStateChange(currState, prevState);
```
Comment (cryptotavares) [inline app/scripts/metamask-controller.js:2533]: Can we be more explicit about `onPreferencesControllerStateChange` name? It seems rather generic and it is not the only place on the where we are subscribing to preferencesController state changes. Above in this file we are doing it as well for TokensRate and CurrencyRate controllers.
Another option would be move all that logic into this one method. But that should be outside of the scope for this PR.

**Review by cryptotavares — DISMISSED** [review]

**Review by vinistevam — DISMISSED** [review]

File: app/scripts/metamask-controller.js (hunk 1 above)
Comment (matthewwalsh0) [reply app/scripts/metamask-controller.js:2533]: Previously we had many small preferences listeners spread around, so the name was intentional as the intent of this method was to create a generic container that would call other more specific methods.

So much of the logic of the `MetamaskController` is just within the constructor so I try to break it down into distinct methods whenever possible, for better readability and maintenance.

**Comment by sleepytanya** [comment]
Token names are displayed when any of the following toggles enabled and not visible when all three toggles are off:

- Autodetect tokens              (Security and Privacy)
- Allow nicknames                 (Experimental)
- Estimate balance changes (Security and Privacy)

**Review by vinistevam — APPROVED** [review]

**Review by cryptotavares — APPROVED** [review]

**Comment by sleepytanya** [comment]
New toggle in the onboarding settings:

Chrome


https://github.com/MetaMask/metamask-extension/assets/104780023/23dbac13-74ee-44e0-92ea-dcff7bdb17ce


Firefox


https://github.com/MetaMask/metamask-extension/assets/104780023/9aa96b18-8a45-40be-89d9-59e3cda056f1

**Pull request merged**